common/random.go

```diff
@@ -45,84 +51,123 @@ MAIN_LOOP:
 }
 
 func RandUint16() uint16 {
-	return uint16(rand.Uint32() & (1<<16 - 1))
+	return uint16(RandUint32() & (1<<16 - 1))
 }
 
 func RandUint32() uint32 {
-	return rand.Uint32()
+	rng.Lock()
+	u32 := rng.Uint32()
+	rng.Unlock()
+	return u32
 }
 
 func RandUint64() uint64 {
-	return uint64(rand.Uint32())<<32 + uint64(rand.Uint32())
+	return uint64(RandUint32())<<32 + uint64(RandUint32())
 }
 
 func RandUint() uint {
-	return uint(rand.Int())
+	rng.Lock()
+	i := rng.Int()
+	rng.Unlock()
+	return uint(i)
 }
 
 func RandInt16() int16 {
-	return int16(rand.Uint32() & (1<<16 - 1))
+	return int16(RandUint32() & (1<<16 - 1))
 }
 
 func RandInt32() int32 {
-	return int32(rand.Uint32())
+	return int32(RandUint32())
 }
 
 func RandInt64() int64 {
-	return int64(rand.Uint32())<<32 + int64(rand.Uint32())
+	return int64(RandUint64())
 }
 
 func RandInt() int {
-	return rand.Int()
+	rng.Lock()
+	i := rng.Int()
+	rng.Unlock()
+	return i
+}
+
+func RandInt31() int32 {
+	rng.Lock()
+	i31 := rng.Int31()
+	rng.Unlock()
+	return i31
+}
+
+func RandInt63() int64 {
+	rng.Lock()
+	i63 := rng.Int63()
+	rng.Unlock()
+	return i63
 }
 
 // Distributed pseudo-exponentially to test for various cases
 func RandUint16Exp() uint16 {
-	bits := rand.Uint32() % 16
+	bits := RandUint32() % 16
 	if bits == 0 {
 		return 0
 	}
 	n := uint16(1 << (bits - 1))
-	n += uint16(rand.Int31()) & ((1 << (bits - 1)) - 1)
+	n += uint16(RandInt31()) & ((1 << (bits - 1)) - 1)
 	return n
 }
 
 // Distributed pseudo-exponentially to test for various cases
 func RandUint32Exp() uint32 {
-	bits := rand.Uint32() % 32
+	bits := RandUint32() % 32
 	if bits == 0 {
 		return 0
 	}
 	n := uint32(1 << (bits - 1))
-	n += uint32(rand.Int31()) & ((1 << (bits - 1)) - 1)
+	n += uint32(RandInt31()) & ((1 << (bits - 1)) - 1)
 	return n
 }
 
 // Distributed pseudo-exponentially to test for various cases
 func RandUint64Exp() uint64 {
-	bits := rand.Uint32() % 64
+	bits := RandUint32() % 64
 	if bits == 0 {
 		return 0
 	}
 	n := uint64(1 << (bits - 1))
-	n += uint64(rand.Int63()) & ((1 << (bits - 1)) - 1)
+	n += uint64(RandInt63()) & ((1 << (bits - 1)) - 1)
 	return n
 }
 
 func RandFloat32() float32 {
-	return rand.Float32()
+	rng.Lock()
+	f32 := rng.Float32()
+	rng.Unlock()
+	return f32
 }
 
 func RandTime() time.Time {
 	return time.Unix(int64(RandUint64Exp()), 0)
 }
 
 func RandBytes(n int) []byte {
-	bs := make([]byte, n)
-	for i := 0; i < n; i++ {
-		bs[i] = byte(rand.Intn(256))
-	}
-	return bs
+	return cRandBytes(n)
```

Comment:
We should keep the old one... cRandBytes is not guaranteed to be fast...

Reply:
See PR comments

Reply:
Gotcha gotcha, that's reasonable.

Reply:
Oh man, great call @jaekwon to ask to revert this. I just created and ran benchmarks and using cRandBytes(n) is between ~3.5X slower!!

Reply:
### Before this change with cRand
```shell
$ go test -run=^$ -bench=.
goos: darwin
goarch: amd64
pkg: github.com/tendermint/tmlibs/common
BenchmarkRandBytes10B-4      	 5000000	       276 ns/op
BenchmarkRandBytes100B-4     	  500000	      2542 ns/op
BenchmarkRandBytes1KiB-4     	   50000	     27274 ns/op
BenchmarkRandBytes100KiB-4   	      50	  27388692 ns/op
BenchmarkRandBytes1MiB-4     	       1	29350694138 ns/op
PASS
```

### After this change
```shell
ok  	github.com/tendermint/tmlibs/common	35.452s
Emmanuels-MacBook-Pro-2:common emmanuelodeke$ go test -run=^$ -bench=.
goos: darwin
goarch: amd64
pkg: github.com/tendermint/tmlibs/common
BenchmarkRandBytes10B-4      	 2000000	       806 ns/op
BenchmarkRandBytes100B-4     	  200000	      7210 ns/op
BenchmarkRandBytes1KiB-4     	   20000	     73042 ns/op
BenchmarkRandBytes100KiB-4   	      20	  79803972 ns/op
BenchmarkRandBytes1MiB-4     	       1	81293929913 ns/op
PASS
ok  	github.com/tendermint/tmlibs/common	89.242s
```
yeezy yeezy!!